File: client2.c
#include <stdio.h>
#include <string.h>
#include <stdint.h>
#include <sys/socket.h>
#include <arpa/inet.h>
#include <unistd.h>
#include <stdlib.h>

/* ===================== PARITY ===================== */
// Calculate even parity
int calculateParity(char *text) {
    int ones = 0;
    for (int i = 0; i < strlen(text); i++) {
        unsigned char c = text[i];
        for (int bit = 7; bit >= 0; bit--) {
            ones += ((c >> bit) & 1);
        }
    }
    return ones % 2; // even parity
}

/* ===================== 2D PARITY ===================== */
void calculate2DParity(char *text, char *rowParity, char *colParity) {
    int len = strlen(text);
    int colOnes[8] = {0};

    for (int i = 0; i < len; i++) {
        unsigned char c = text[i];
        int rowOnes = 0;

        for (int bit = 7; bit >= 0; bit--) {
            int b = (c >> bit) & 1;
            rowOnes += b;
            colOnes[7 - bit] += b;
        }
        rowParity[i] = (rowOnes % 2) + '0';
    }
    rowParity[len] = '\0';

    for (int j = 0; j < 8; j++)
        colParity[j] = (colOnes[j] % 2) + '0';

    colParity[8] = '\0';
}

/* ===================== CRC-8 ===================== */
unsigned char crc8(unsigned char *data, int len) {
    unsigned char crc = 0;

    for (int i = 0; i < len; i++) {
        crc ^= data[i];
        for (int j = 0; j < 8; j++) {
            if (crc & 0x80)
                crc = (crc << 1) ^ 0x07;
            else
                crc <<= 1;
        }
    }
    return crc;
}

/* ===================== INTERNET CHECKSUM ===================== */
uint16_t internetChecksum(unsigned char *data, int len) {
    uint32_t sum = 0;

    for (int i = 0; i < len; i += 2) {
        uint16_t word = data[i] << 8;
        if (i + 1 < len)
            word |= data[i + 1];

        sum += word;
        if (sum > 0xFFFF)
            sum = (sum & 0xFFFF) + 1;
    }
    return ~sum;
}

int main() {
    printf("Client 2 is running (Receiver)...\n");

    int sock = socket(AF_INET, SOCK_STREAM, 0);
    if (sock < 0) {
        printf("Failed to create socket!\n");
        return 1;
    }

    struct sockaddr_in server;
    server.sin_family = AF_INET;
    server.sin_port = htons(8081);
    server.sin_addr.s_addr = inet_addr("127.0.0.1");

    if (connect(sock, (struct sockaddr*)&server, sizeof(server)) < 0) {
        printf("Failed to connect to server!\n");
        return 1;
    }

    printf("Connected to server. Waiting for packets...\n");

    char buffer[8192];

    while (1) {
        int received = recv(sock, buffer, sizeof(buffer) - 1, 0);
        if (received <= 0) {
            printf("Server closed the connection!\n");
            break;
        }

        buffer[received] = '\0';
        printf("\n--- Packet Received ---\n%s\n", buffer);

        // Split packet
        char *data = strtok(buffer, "|");
        char *method = strtok(NULL, "|");
        char *control = strtok(NULL, "|");

        if (!data || !method || !control) {
            printf("INVALID PACKET FORMAT!\n");
            continue;
        }

        printf("DATA: %s\n", data);
        printf("METHOD: %s\n", method);
        printf("CONTROL: %s\n", control);

        /* ===================== VERIFICATION ===================== */

        if (strcmp(method, "PARITY") == 0) {
            int parity = calculateParity(data);
            printf("Calculated Parity = %d\n", parity);
            printf((parity == atoi(control)) ? "NO ERROR DETECTED\n" : "ERROR DETECTED\n");
        }

        else if (strcmp(method, "2DPARITY") == 0) {
            char row[256], col[9];
            calculate2DParity(data, row, col);

            printf("Calculated Row Parity: %s\n", row);
            printf("Calculated Column Parity: %s\n", col);

            char *recvRow = control;
            char *recvCol = strtok(NULL, "|");

            if (!recvCol) {
                printf("PACKET ERROR: Invalid 2DPARITY format!\n");
                continue;
            }

            int ok = (!strcmp(row, recvRow) && !strcmp(col, recvCol));
            printf(ok ? "NO ERROR DETECTED\n" : "ERROR DETECTED\n");
        }

        else if (strcmp(method, "CRC8") == 0) {
            unsigned char crc = crc8((unsigned char*)data, strlen(data));
            printf("Calculated CRC8 = %02X\n", crc);
            printf((crc == strtol(control, NULL, 16)) ? "NO ERROR DETECTED\n" : "ERROR DETECTED\n");
        }

        else if (strcmp(method, "IPCHECKSUM") == 0) {
            uint16_t checksum = internetChecksum((unsigned char*)data, strlen(data));
            printf("Calculated IP Checksum = %04X\n", checksum);
            printf((checksum == strtol(control, NULL, 16)) ? "NO ERROR DETECTED\n" : "ERROR DETECTED\n");
        }

        else if (strcmp(method, "HAMMING") == 0) {
            printf("Starting Hamming decode...\n");

            if (strlen(data) != 7) {
                printf("ERROR: Hamming(7,4) requires 7-bit input!\n");
            } else {
                int b1 = data[0] - '0';
                int b2 = data[1] - '0';
                int b3 = data[2] - '0';
                int b4 = data[3] - '0';
                int b5 = data[4] - '0';
                int b6 = data[5] - '0';
                int b7 = data[6] - '0';

                int p1 = b1 ^ b3 ^ b5 ^ b7;
                int p2 = b2 ^ b3 ^ b6 ^ b7;
                int p3 = b4 ^ b5 ^ b6 ^ b7;

                int errorBit = (p3 << 2) | (p2 << 1) | p1;

                if (errorBit == 0) {
                    printf("NO ERROR DETECTED. Data is correct.\n");
                } else {
                    printf("ERROR DETECTED! Faulty bit position: %d\n", errorBit);

                    int index = errorBit - 1;
                    char corrected[8];
                    strcpy(corrected, data);
                    corrected[index] = (corrected[index] == '0') ? '1' : '0';

                    printf("Corrected Data: %s\n", corrected);
                }
            }
        }

        else {
            printf("UNKNOWN METHOD!\n");
        }
    }

    close(sock);
    return 0;
}

File: client2.py
# client2.py
# Client 2 – Receiver + Error Controller

import socket

HOST = "127.0.0.1"
PORT = 6000

# ---------------- Utility Functions ----------------

def text_to_binary(text):
    return ''.join(format(ord(c), '08b') for c in text)

# Parity (Even / Odd)
def parity_bit(binary, mode="even"):
    ones = binary.count("1")
    if mode == "even":
        parity = ones % 2
    else:
        parity = (ones + 1) % 2
    return binary + str(parity)

# 2D Parity
def parity_2d(binary, row_size=8):
    rows = [binary[i:i+row_size] for i in range(0, len(binary), row_size)]
    matrix = []

    for row in rows:
        matrix.append(row + str(row.count("1") % 2))

    col_parity = ""
    for col in range(len(matrix[0])):
        col_parity += str(sum(int(r[col]) for r in matrix) % 2)

    matrix.append(col_parity)
    return '\n'.join(matrix)

# CRC-16 (CCITT)
def crc16(data):
    crc = 0xFFFF
    for b in data.encode():
        crc ^= b << 8
        for _ in range(8):
            crc = (crc << 1) ^ 0x1021 if crc & 0x8000 else crc << 1
            crc &= 0xFFFF
    return format(crc, "04X")

# Hamming (7,4)
def hamming_encode(binary):
    encoded = ""
    for i in range(0, len(binary), 4):
        d = binary[i:i+4].ljust(4, '0')
        d1, d2, d3, d4 = map(int, d)

        p1 = d1 ^ d2 ^ d4
        p2 = d1 ^ d3 ^ d4
        p3 = d2 ^ d3 ^ d4

        encoded += f"{p1}{p2}{d1}{p3}{d2}{d3}{d4}"
    return encoded

# Internet Checksum
def internet_checksum(data):
    if len(data) % 2 != 0:
        data += '\0'

    checksum = 0
    for i in range(0, len(data), 2):
        word = (ord(data[i]) << 8) + ord(data[i+1])
        checksum += word
        checksum = (checksum & 0xFFFF) + (checksum >> 16)

    return format(~checksum & 0xFFFF, "04X")

# ---------------- Receive Packet ----------------

client = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
client.connect((HOST, PORT))

packet = client.recv(8192).decode()
client.close()

print("\nReceived packet:")
print(packet)

# ---------------- Partition Packet ----------------

try:
    data, method, incoming_control = packet.split("|", 2)
except ValueError:
    print("Invalid packet format")
    exit()

binary = text_to_binary(data)
computed_control = ""

# ---------------- Re-compute Control ----------------

if method == "PARITY_EVEN":
    computed_control = parity_bit(binary, "even")

elif method == "PARITY_ODD":
    computed_control = parity_bit(binary, "odd")

elif method == "2D_PARITY":
    computed_control = parity_2d(binary)

elif method == "CRC16":
    computed_control = crc16(data)

elif method == "HAMMING":
    computed_control = hamming_encode(binary)

elif method == "CHECKSUM":
    computed_control = internet_checksum(data)

else:
    print("Unknown method")
    exit()

# ---------------- Compare & Output ----------------

status = "DATA CORRECT" if computed_control == incoming_control else "DATA CORRUPTED"

print("\n---------------- RESULT ----------------")
print(f"Received Data        : {data}")
print(f"Method               : {method}")
print(f"Sent Check Bits      : {incoming_control}")
print(f"Computed Check Bits  : {computed_control}")
print(f"Status               : {status}")
print("---------------------------------------")
